### PR TITLE
Fix most_recent_version for non Ruby Platform Gems

### LINKED ISSUE #6348: Gem page not updating
## Steps to Reproduce
1. Visit https://rubygems.org/gems/getargv
2. note latest version in Versions: list at bottom of page
3. note version next to header and in install instructions
4. notice they don't match

## Expected Behavior
Gem page should update in a reasonable amount of time

## Current Behavior
Page hasn't updated after approx 4.5 hours

## Environment
Browser and its version: FF 149.0b10
RubyGems version: N/A

### PR DESCRIPTION
**Pull request by girachawda**
## What

- Closes: https://github.com/rubygems/rubygems.org/issues/6348

Fix `most_recent_version` returning the wrong version for gems that only have non-ruby platform versions (e.g. `universal-darwin`).

## Why

The `most_recent_version` association used a `CASE WHEN latest THEN number ELSE NULL END DESC` ordering. 

**PostgreSQL sorts `NULL` first in `DESC` order**, so versions with `latest=false` (producing `NULL`) were incorrectly preferred over versions with `latest=true`. For gems with only non-ruby platform versions, the first ordering clause is a tie (all score 1), so the broken second clause decided the result — picking the wrong version.

The fix replaces `NULL`-producing expressions with explicit `0`/`1` integers and uses the `position` column (which is already correctly computed via `Gem::Version` semantic comparison in `reorder_versions`.

## Tophat

I created a mock gem page with versions locally and confirmed that header shows the correct version
Monosnap testgem | RubyGems org | your community gem host 2026-05-06 14-33-14

You can also run tests

```bash
bin/rails test test/models/rubygem_test.rb -n "/most recent non-ruby/"
```

### CODE REVIEW

File: app/models/rubygem.rb [hunk 1]
```diff
@@ -60,9 +60,11 @@ def unique_reverse_runtime_dependencies
...
     lambda {
-      order(Arel.sql("case when #{quoted_table_name}.latest AND #{quoted_table_name}.platform = 'ruby' then 2 else 1 end desc"))
-        .order(Arel.sql("case when #{quoted_table_name}.latest then #{quoted_table_name}.number else NULL end desc"))
-        .order(id: :desc)
+      order(
+        Arel.sql("case when #{quoted_table_name}.latest AND #{quoted_table_name}.platform = 'ruby' then 0 else 1 end"),
+        Arel.sql("case when #{quoted_table_name}.latest then 0 else 1 end"),
+        :position
+      ).order(id: :desc)
     },
```
Comment (jenshenny) [inline app/models/rubygem.rb:69]: I'm finding the 0s and 1s hard to track 😅 , wdyt about this? I think it's easy to see what's higher priority. Curious to hear if you were deciding between different queries that would fix the issue.

```suggestion
      order(
        Arel.sql("case when #{quoted_table_name}.latest AND #{quoted_table_name}.platform = 'ruby' then 0 " \
                 "when #{quoted_table_name}.latest then 1 else 2 end"),
        :position,
        id: :desc
      )
    },
```

**Review by jenshenny — COMMENTED** [review]
Thanks for working on this!

File: test/models/rubygem_test.rb [hunk 2]
```diff
@@ -163,6 +163,15 @@ class RubygemTest < ActiveSupport::TestCase
       assert_equal version1pre, @rubygem.reload.most_recent_version
     end
 
+    should "return the most recent non-ruby platform version when multiple non-ruby platform versions exist" do
```
Comment (colby-swandale) [inline test/models/rubygem_test.rb:166]: This test doesn't exercise the bug. 0.3.18 vs 0.3.19 happens to lex-sort the same as semver-sort.

Walking the old ordering against this test data:
- case when latest AND platform='ruby' then 2 else 1 end DESC → both rows score 1 (tie)
- case when latest then number else NULL end DESC → Postgres compares the two number strings. Lexicographically '0.3.19' > '0.3.18', so version2 wins.

So the old code passes this test too. The bug only manifests when lex sort and `Gem::Version` sort disagree.

```suggestion
    should "return the most recent non-ruby platform version when multiple non-ruby platform versions exist" do
      create(:version, rubygem: @rubygem, number: "0.3.9", platform: "universal-darwin")
      version2 = create(:version, rubygem: @rubygem, number: "0.3.10", platform: "universal-darwin")

      @rubygem.reorder_versions

      assert_equal version2, @rubygem.reload.most_recent_version
    end
```

File: app/models/rubygem.rb (hunk 1 above)
Comment (girachawda) [reply app/models/rubygem.rb:69]: Fixed this!

File: test/models/rubygem_test.rb (hunk 2 above)
Comment (girachawda) [reply test/models/rubygem_test.rb:166]: Hey! Thanks for the review. I updated the version numbers to 0.3.9/0.3.10. Good call on making lex-sort and semver-sort disagree. 

One thing I noticed though, when I reverted the fix and ran the original test (with 0.3.18/0.3.19), it actually still failed against the old code. I think it's because `reorder_versions` only marks one version per platform as `latest=true`, so the CASE expression ends up comparing `"0.3.19"` vs `NULL` rather than two number strings. 

And PostgreSQL sorts NULLs first in DESC, so it picks the wrong version. I also added a second test with different platforms (x86_64-linux + universal-darwin) to cover the string comparison case where both versions have `latest=true`. Let me know if I'm missing something!

Comment (jenshenny) [reply test/models/rubygem_test.rb:166]: I agree with @girachawda's read here. The original test did exercise the bug because `reorder_versions` only marks one version per platform as `latest=true` but still chooses the wrong version because of the null sorting. It does match with what the original bug report mentioned with `0.3.18/0.3.19` versions.

👍 to adding another test to catch lex and semver sort since that issue exists too as @colby-swandale brought up.

File: test/models/rubygem_test.rb [hunk 3]
```diff
@@ -163,6 +163,24 @@ class RubygemTest < ActiveSupport::TestCase
...
+      create(:version, rubygem: @rubygem, number: "0.3.9", platform: "universal-darwin")
+      version2 = create(:version, rubygem: @rubygem, number: "0.3.10", platform: "universal-darwin")
```
Comment (jenshenny) [inline test/models/rubygem_test.rb]:
```suggestion
      create(:version, rubygem: @rubygem, number: "0.3.8", platform: "universal-darwin")
      version2 = create(:version, rubygem: @rubygem, number: "0.3.9", platform: "universal-darwin")
```

nit: It looks like we are testing both the number lex ordering and NULLs first case in this first test.

File: test/models/rubygem_test.rb [hunk 4]
```diff
@@ -163,6 +163,24 @@ class RubygemTest < ActiveSupport::TestCase
...
+
+      @rubygem.reorder_versions
+
+      assert_equal version2, @rubygem.reload.most_recent_version
+    end
+
+    should "return the highest version across platforms when lex-sort and semver-sort disagree" do
```
Comment (jenshenny) [inline test/models/rubygem_test.rb]:
```suggestion
    should "return the highest semver version across platforms" do
```

File: test/models/rubygem_test.rb (hunk 3 above)
Comment (girachawda) [reply test/models/rubygem_test.rb]: good catch!

File: test/models/rubygem_test.rb [hunk 5]
```diff
@@ -163,6 +163,24 @@ class RubygemTest < ActiveSupport::TestCase
...
+
+    should "return the highest semver version across platforms" do
+      create(:version, rubygem: @rubygem, number: "0.3.9", platform: "x86_64-linux")
+      version2 = create(:version, rubygem: @rubygem, number: "0.3.10", platform: "universal-darwin")
+
+      @rubygem.reorder_versions
+
+      assert_equal version2, @rubygem.reload.most_recent_version
+    end
+
```
Comment (OughtPuts) [inline test/models/rubygem_test.rb:183]: Minor suggestion, non-blocking:

We could decide to cover a few more edge cases in the tests as this is an important scope.

1. Ruby vs non-ruby priority with higher non-ruby version: where a ruby platform has v1.0.0 (latest) and a non-ruby platform has v2.0.0 (latest), asserting that the ruby version is still preferred. 

2. Multiple non-ruby platforms both marked latest: e.g. 0.4.0 on x86_64-linux and 0.3.9 on universal-darwin, both latest for their platform, asserting the higher version wins.

**Review by OughtPuts — COMMENTED** [review]
Looks good! I can't approve or request changes officially but wanted to read the PR anyway and leave a thought. I think your work on this is really good though! 🔥

File: test/models/rubygem_test.rb [hunk 6]
```diff
@@ -163,6 +163,33 @@ class RubygemTest < ActiveSupport::TestCase
...
+    should "return the highest version when multiple non-ruby platforms are both marked latest" do
+      create(:version, rubygem: @rubygem, number: "0.3.9", platform: "universal-darwin")
+      version2 = create(:version, rubygem: @rubygem, number: "0.4.0", platform: "x86_64-linux")
+
+      @rubygem.reorder_versions
+
+      assert_equal version2, @rubygem.reload.most_recent_version
+    end
```
Comment (jenshenny) [inline test/models/rubygem_test.rb]: This test is testing the same case as `return the highest semver version across platforms` if I understand correctly. They both test that if there's two different non ruby platforms that the higher semver is chosen.

Comment (girachawda) [reply test/models/rubygem_test.rb]: Fixed!

**Review by jenshenny — APPROVED** [review]
